### PR TITLE
validation: fix set readonly variable max_allowed_packet should return error

### LINKED ISSUE #32349: `set @@max_allowed_packet=default;` should return error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
set @@max_allowed_packet=default;
```

### 2. What did you expect to see? (Required)
```
mysql> set @@max_allowed_packet=default;
ERROR 1621 (HY000): SESSION variable 'max_allowed_packet' is read-only. Use SET GLOBAL to assign the value
```

### 3. What did you see instead (Required)
```
mysql> set @@max_allowed_packet=default;
Query OK, 0 rows affected (0.00 sec)
```

### PR DESCRIPTION
**Pull request by fanrenhoo**
Signed-off-by: fanrenhoo <fanrenhoo@sina.com>

### What problem does this PR solve?

Issue Number: close #32349

Problem Summary: when set read only variable max_allowed_packet should return error, but current code no error return

### What is changed and how it works?
in logic planner check if its read only variable and not global set statement, so return error as required, compatible with mysql

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or st
mysql> set @@max_allowed_packet=default;
ERROR 1621 (HY000): SESSION variable 'max_allowed_packet' is read-only. Use SET GLOBAL to assign the valueeps below)

mysql> set @@global.max_allowed_packet=default;
Query OK, 0 rows affected (0.00 sec)

- [ ] No code

Side effects
No
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation
No
- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Make the set read only variable max_allowed_packet return error and mysql does, fix the issue #32349
```

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -834,6 +834,10 @@ func (b *PlanBuilder) buildSet(ctx context.Context, v *ast.SetStmt) (Plan, error
 			err := ErrSpecificAccessDenied.GenWithStackByArgs("RESTRICTED_VARIABLES_ADMIN")
 			b.visitInfo = appendDynamicVisitInfo(b.visitInfo, "RESTRICTED_VARIABLES_ADMIN", false, err)
 		}
+		if !vars.IsGlobal && vars.Name == "max_allowed_packet" {
```
Comment (Defined2014) [inline planner/core/planbuilder.go:837]: You can add a `Validation` func for `max_allowed_packet`, which is a common way to solve this kind of problem.

Comment (fanrenhoo) [reply planner/core/planbuilder.go]: THKS, I made a new fix follow suggestion, please check. BR

Comment (Defined2014) [reply planner/core/planbuilder.go]: Maybe you misunderstand my words. You can add a `Validation` function like this https://github.com/pingcap/tidb/blob/master/sessionctx/variable/sysvar.go#L968

Comment (Defined2014) [reply planner/core/planbuilder.go]: Maybe you misunderstand my words. You can add a `Validation` function like this https://github.com/pingcap/tidb/blob/master/sessionctx/variable/sysvar.go#L968

Comment (fanrenhoo) [reply planner/core/planbuilder.go]: THKS, sorry for misunderstand. I did try in sysVar.go to do sth before, but every time when the server started error return with max_allowed_packet. I'll learn and try again.

Comment (Defined2014) [reply planner/core/planbuilder.go]: Maybe caused by https://github.com/pingcap/tidb/blob/master/session/session.go#L1214. So there are two ways 
* use some way to figure out it's an intertnal session or not in `Validation` function.
* fix it in planner like this PR.

Both OK for me, do you have some comments? @morgo

Comment (fanrenhoo) [reply planner/core/planbuilder.go]: I fixed it in Validation as suggested in internal session case. please check, THKS much

Comment (morgo) [reply planner/core/planbuilder.go]: Nice trick to see if it is a user session! LGTM

**Review by Defined2014 — APPROVED** [review]

**Comment by Defined2014** [comment]
/cc @morgo @bb7133
PTAL

**Comment by fanrenhoo** [comment]
Hi, how about this one, THKS

**Comment by fanrenhoo** [comment]
Hi, @morgo @bb7133 how about this one, THKS

**Comment by morgo** [comment]
> Hi, @morgo @bb7133 how about this one, THKS

LGTM, let me just check tests still pass against master. We recently added more correct handling of max_allowed_packet, want to make sure they still work.

**Review by morgo — APPROVED** [review]

**Pull request merged**